### ISSUE TITLE
v2 visibilities are the complex conjugate of v4's visibilities

### ISSUE THREAD

**Issue by sjperkins** [ska-sa/montblanc#84]
v4's visibilities are the complex conjugate of v2's.

There'll be an incorrect sign somewhere. First place to look is the [jones_multiply_4x4_hermitian_transpose_in_place](https://github.com/ska-sa/montblanc/blob/0a4e7a810a722401842a56c04446312c58165ddf/montblanc/src/montblanc/include/jones.cuh#L74-L114) function. Inverting the sign of the imaginary part of the complex number produces the correct result for simple cases (default [1 0 0 1] E and G terms).

As reported below by @saiyanprince , this effects the BIRO solver in the following manner:

> We first noticed this behaviour when v2 and v4 were not converging to the same values for the position estimates (l,m) during the Bayesian analysis. Increasing the prior boxes to include both (-l,-m) and (l,m), it became apparent that a source at (l,m) (as interpreted by v2) converged to (-l,-m) in v4.

A temporary workaround is to invert the **(l,m)** coordinates found by the BIRO solver. However, v4 does not agree with MeqTrees (while v2 does).

**Comment by sjperkins** [comment]
@saiyanprince I've created a placeholder issue for this. Mail me or comment on this issue if you would like the above updated.

**Comment by saiyanprince** [comment]
We first noticed this behaviour when v2 and v4 were not converging to the same values for the position estimates (l,m) during the Bayesian analysis. Increasing the prior boxes to include both (-l,-m) and (l,m), it became apparent that a source at (l,m) (as interpreted by v2) converged to (-l,-m) in v4.

@sjperkins @jtlz2 From our previous discussion, the (u,v,w) co-ordinates are fine. I'll make this modification and test it on the position and shape estimates.

**Comment by marziarivi** [comment]
I have some problems with the positions fitting too. By the way, they are assumed to be measured in arcsec, aren't they?

**Comment by jtlz2** [comment]
@marziarivi radians away from pointing centre, no?

**Comment by marziarivi** [comment]
Ah, so they are measured in radiants... ok.

**Comment by sjperkins** [comment]
@marziarivi @jtlz2 I think we've been working with arcseconds in general.

In general, I think its fair to say that if you're working with MeasurementSets, then the units of the MS inform the rest of the input that the user supplies to Montblanc.

**Comment by saiyanprince** [comment]
Any size parameter, like position or shape, is input to montblanc in
radians, which are also the units in which Measurement Sets record them.

In our analyses, the sampling is done in radians. It's just that the user
is allowed to set the prior ranges in arcseconds.

**Comment by marziarivi** [comment]
Thanks for the clarification.

**Comment by marziarivi** [comment]
So, for gaussian sources are you assuming in your model that el and em parameters are provided in radiants (even if the order of magnitude are arcsecs)?

**Comment by o-smirnov** [comment]
In my opinion, this is correct. An API should stick to consistent units
throughout, and preferably SI units. So all angles should be radians.

>
Conversion to more human units, e.g. arcsecs for source extents, if
necessary, should happen in the user-facing layer.

**Comment by marziarivi** [comment]
Ok. Thanks. Just to be sure sersic model units be consistent with the rest of the code.

**Comment by sjperkins** [comment]
Problem is likely because order of multiplication is incorrect and subtraction is incorrect. [Here](https://github.com/ska-sa/montblanc/blob/0641b53ecd491117e0c1353e39ed50ace9b43453/montblanc/impl/biro/v4/cpu/SolverCPU.py#L45)

```python
u = ne.evaluate('ap-aq', {'ap': u[1], 'aq': u[0]})
```

should be

```python
u = ne.evaluate('aq-ap', {'aq': u[1], 'ap': u[0]})
```

and [here](https://github.com/ska-sa/montblanc/blob/0641b53ecd491117e0c1353e39ed50ace9b43453/montblanc/impl/biro/v4/gpu/RimeSumCoherencies.py#L180)

```python
montblanc::jones_multiply_4x4_hermitian_transpose_in_place<T>(ant_two, ant_one);
```

should be

```python
montblanc::jones_multiply_4x4_hermitian_transpose_in_place<T>(ant_one, ant_two);
```

**Comment by marziarivi** [comment]
On 4 Mar 2016, at 07:56, Simon Perkins <notifications@github.com> wrote:

> Problem is likely because order of multiplication is incorrect and subtraction is incorrect. Here
> 
> u = ne.evaluate('ap-aq', {'ap': u[1], 'aq': u[0]})
> should be
> 
> u = ne.evaluate('aq-ap', {'aq': u[1], 'ap': u[0]})

I don’t see any difference between the two: you inverted the order of subtraction and the name of the variables, so at the end you perform the same operation.

> and here
> 
> montblanc::jones_multiply_4x4_hermitian_transpose_in_place<T>(ant_two, ant_one);
> should be
> 
> montblanc::jones_multiply_4x4_hermitian_transpose_in_place<T>(ant_one, ant_two);
Does this mean that in the next lines of the code we have to use ant_one instead of ant_two?

**Comment by sjperkins** [comment]
@marziarivi 

> I don’t see any difference between the two: you inverted the order of subtraction and the name of the variables, so at the end you perform the same operation.

Yes, it was a typo on my part, I've updated my original comment.

> Does this mean that in the next lines of the code we have to use ant_one instead of ant_two?

Yes, thanks for spotting that!

**Comment by marziarivi** [comment]
On 7 Mar 2016, at 09:53, Simon Perkins <notifications@github.com> wrote:

> @marziarivi
> 
> I don’t see any difference between the two: you inverted the order of subtraction and the name of the variables, so at the end you perform the same operation.
> 
> Yes, it was a typo on my part, I've updated my original comment.
> 
Why this coordinate should be treated differently from the others? From what I see you compute the baselines as follows: u[0]-u[1], v[1]-v[0], w[1]-w[0].

**Comment by marziarivi** [comment]
I agree the problem is in the call of

montblanc::jones_multiply_4x4_hermitian_transpose_in_place<T>(ant_two, ant_one);

as you said. Since it is called several times and you have to replace ant_two with ant_one after, wouldn't  be simpler to modify the function jones_multiply_4x4_hermitian_transpose_in_place only?

**Comment by sjperkins** [comment]
Closed by #93
